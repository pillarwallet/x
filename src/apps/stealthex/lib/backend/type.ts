export type UnavailableResponse = {
    state: 'unknown';
  };
  
export type FailedResponse<T> = {
    state: 'failed';
    status: number;
    message?: T;
};

export type ThrowableError<T> = UnavailableResponse | FailedResponse<T>;

export type ErrorResponse<T> = ThrowableError<T> & { data: null };

export type ThrowableResponse<T> = {
    status: number;
    data: T;
    headers: Headers;
};

export type SuccessResponse<T> = ThrowableResponse<T> & {
state: 'success';
};
  
export type Response<D, E> = SuccessResponse<D> | ErrorResponse<E>;

export type RequestResponse<
  T extends boolean,
  D = unknown,
  E = string,
> = T extends true ? ThrowableResponse<D> : Response<D, E>;