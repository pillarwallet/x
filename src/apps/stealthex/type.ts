import { CurrencyData } from './lib/backend/api';

export type Config = {
    amount: string;
    fixed: boolean;
    fiat: boolean;
    sendCurrency: CurrencyData;
    receiveCurrency: CurrencyData;
    getPairs?: CurrencyData[];
    sendPairs?: CurrencyData[];
    recepientAddress?: string;
    lockReceiveCurrency?: boolean;
}

export type ExchangeInfoCurrencyPair = {
    [key: string]: CurrencyData | undefined;
  };

export type ExchangeInfo = {
    id: string;
    type: string;
    timestamp: string;
    updated_at: string;
    currency_from: string;
    currency_to: string;
    amount_from: string;
    expected_amount: string;
    amount_to: string;
    amount_to_btc: string | null;
    address_from: string;
    address_to: string;
    extra_id_from: string | null;
    extra_id_to: string | null;
    tx_from: string | null;
    tx_to: string | null;
    status:
      | 'waiting'
      | 'confirming'
      | 'exchanging'
      | 'sending'
      | 'finished'
      | 'failed'
      | 'refunded'
      | 'expired'
      | 'verifying';
    refund_address: string | null;
    refund_extra_id: string | null;
    currencies: ExchangeInfoCurrencyPair;
    review?: boolean;
    payment_id?: string;
    mercuryo_signature: string;
};

export type StructuredCurrency = {
  fiat: CurrencyData[];
  popular: CurrencyData[];
  other: CurrencyData[];
};