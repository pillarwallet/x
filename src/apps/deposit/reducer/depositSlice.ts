/* eslint-disable no-param-reassign */
import { createSlice, PayloadAction } from '@reduxjs/toolkit';

// types
import { AddedAssets, BalanceInfo } from '../types/types';

export type DepositState = {
  depositStep: 'list' | 'send';
  selectedAsset: BalanceInfo | AddedAssets | undefined;
};

const initialState: DepositState = {
  depositStep: 'list',
  selectedAsset: undefined,
};

const depositSlice = createSlice({
  name: 'deposit',
  initialState,
  reducers: {
    setDepositStep(state, action: PayloadAction<'list' | 'send'>) {
      state.depositStep = action.payload;
    },
    setSelectedAsset(
      state,
      action: PayloadAction<BalanceInfo | AddedAssets | undefined>
    ) {
      state.selectedAsset = action.payload;
    },
  },
});

export const { setDepositStep, setSelectedAsset } = depositSlice.actions;

export default depositSlice;
