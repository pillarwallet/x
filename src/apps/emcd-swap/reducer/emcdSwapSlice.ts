import { createSlice, PayloadAction } from '@reduxjs/toolkit';

import { VIEW_TYPE } from '../constants/views';

interface ICoin {
  title: string;
  icon_url: string;
  networks: Array<Record<string, string>>
}

export type SwapFormData = {
  address_to?: string | null
  tag_to?: string | null
  email?: string | null
  amount_from?: string | null
  amount_to?: string | null
  coin_from?: string | null
  coin_to?: string | null
  network_from?: string | null
  network_to?: string | null
}

export type DetailSwapFormData = {
  address_from: string | null;
  address_to: string | null;
  amount_from: string | null;
  amount_to: string | null;
  coin_from: string | null;
  coin_to: string | null;
  network_from: string | null;
  network_to: string | null;
  rate: string | null;
  start_time: number | null;
  status: number | null;
  swap_duration: number | null;
  tag_from: string | null;
  tag_to: string | null;
}

export type EmcdSwapState = {
  swapID: string | null,
  depositAddress: string | null,
  currentView: VIEW_TYPE,
  backView: VIEW_TYPE,
  coins: ICoin[],
  formData: SwapFormData,
  detailSwapFormData: DetailSwapFormData,
}

const initialState: EmcdSwapState = {
  swapID: null,
  depositAddress: null,
  currentView: VIEW_TYPE.EXCHANGE,
  backView: VIEW_TYPE.EXCHANGE,
  coins: [],
  formData: {},
  detailSwapFormData: {
    address_from: null,
    address_to: null,
    amount_from: null,
    amount_to: null,
    coin_from: null,
    coin_to: null,
    network_from: null,
    network_to: null,
    rate: null,
    start_time: null,
    status: null,
    swap_duration: null,
    tag_from: null,
    tag_to: null,
  }
}

const emcdSwapSlice = createSlice({
  name: 'emcdSwap',
  initialState,
  reducers: {
    setCurrentView(state, action: PayloadAction<VIEW_TYPE>) {
      state.backView = state.currentView
      state.currentView = action.payload
    },

    setBackView(state) {
      state.currentView = state.backView
    },

    setCoins(state, action: PayloadAction<ICoin[]>) {
      state.coins = action.payload;
    },

    setSwapID(state, action: PayloadAction<string | null>) {
      state.swapID = action.payload;
    },

    setDepositAddress(state, action: PayloadAction<string | null>) {
      state.depositAddress = action.payload;
    },

    setFormData(state, action: PayloadAction<Partial<SwapFormData>>) {
      state.formData = {
        ...state.formData,
        ...action.payload,
      }
    },

    setDetailSwapFormData(state, action: PayloadAction<DetailSwapFormData>) {
      state.detailSwapFormData = action.payload
    },

    setDetailSwapStatus(state, action: PayloadAction<number>) {
      state.detailSwapFormData.status = action.payload
    }
  }
})

export const { setCurrentView, setCoins, setFormData, setSwapID, setDetailSwapFormData, setDepositAddress, setBackView, setDetailSwapStatus } = emcdSwapSlice.actions

// Геттер
export const selectCurrentView = (state: { emcdSwap: EmcdSwapState }) =>
  state.emcdSwap.currentView

export const selectCoins = (state: { emcdSwap: EmcdSwapState }) =>
  state.emcdSwap.coins;

export const selectFormData = (state: { emcdSwap: EmcdSwapState }) =>
  state.emcdSwap.formData;

export const selectDetailFormDataSwap = (state: { emcdSwap: EmcdSwapState }) =>
  state.emcdSwap.detailSwapFormData;

export const selectSwapID = (state: { emcdSwap: EmcdSwapState }) =>
  state.emcdSwap.swapID;

export const selectDepositAddress = (state: { emcdSwap: EmcdSwapState }) =>
  state.emcdSwap.depositAddress;

export const selectDetailSwapStatus = (state: { emcdSwap: EmcdSwapState }) =>
  state.emcdSwap.detailSwapFormData.status;

export default emcdSwapSlice