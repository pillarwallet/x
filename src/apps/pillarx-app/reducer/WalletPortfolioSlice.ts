/* eslint-disable no-param-reassign */
import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { sub } from 'date-fns';

// types
import {
  PortfolioData,
  TokenPriceGraphPeriod,
  WalletHistory,
} from '../../../types/api';

// utils
import { convertDateToUnixTimestamp } from '../../../utils/common';
import { PeriodFilterBalance, PeriodFilterPnl } from '../utils/portfolio';

export type WalletPortfolioState = {
  walletPortfolio: PortfolioData | undefined;
  isWalletPorfolioLoading: boolean;
  isWalletPortfolioErroring: boolean;
  walletPortfolioWithPnl: PortfolioData | undefined;
  isWalletPorfolioWithPnlLoading: boolean;
  isWalletPortfolioWithPnlErroring: boolean;
  priceGraphPeriod: TokenPriceGraphPeriod;
  periodFilter: PeriodFilterBalance;
  periodFilterPnl: PeriodFilterPnl;
  walletHistoryGraph: WalletHistory | undefined;
  isWalletHistoryGraphLoading: boolean;
  isWalletHistoryGraphErroring: boolean;
  topTokenUnrealizedPnL: WalletHistory | undefined;
  isTopTokenUnrealizedPnLLoading: boolean;
  isTopTokenUnrealizedPnLErroring: boolean;
  selectedBalanceOrPnl: 'balance' | 'pnl';
  isRefreshAll: boolean;
  isReceiveModalOpen: boolean;
};

const initialState: WalletPortfolioState = {
  walletPortfolio: undefined,
  isWalletPorfolioLoading: false,
  isWalletPortfolioErroring: false,
  walletPortfolioWithPnl: undefined,
  isWalletPorfolioWithPnlLoading: false,
  isWalletPortfolioWithPnlErroring: false,
  priceGraphPeriod: {
    from: convertDateToUnixTimestamp(sub(new Date(), { days: 1 })),
    to: undefined,
  },
  periodFilter: PeriodFilterBalance.DAY,
  periodFilterPnl: PeriodFilterPnl.DAY,
  walletHistoryGraph: undefined,
  isWalletHistoryGraphLoading: false,
  isWalletHistoryGraphErroring: false,
  topTokenUnrealizedPnL: undefined,
  isTopTokenUnrealizedPnLLoading: false,
  isTopTokenUnrealizedPnLErroring: false,
  selectedBalanceOrPnl: 'balance',
  isRefreshAll: false,
  isReceiveModalOpen: false,
};

const walletPortfolioSlice = createSlice({
  name: 'walletPortfolio',
  initialState,
  reducers: {
    setWalletPortfolio(
      state,
      action: PayloadAction<PortfolioData | undefined>
    ) {
      state.walletPortfolio = action.payload;
    },
    setIsWalletPorfolioLoading(state, action: PayloadAction<boolean>) {
      state.isWalletPorfolioLoading = action.payload;
    },
    setIsWalletPortfolioErroring(state, action: PayloadAction<boolean>) {
      state.isWalletPortfolioErroring = action.payload;
    },
    setWalletPortfolioWithPnl(
      state,
      action: PayloadAction<PortfolioData | undefined>
    ) {
      state.walletPortfolioWithPnl = action.payload;
    },
    setIsWalletPortfolioWithPnlLoading(state, action: PayloadAction<boolean>) {
      state.isWalletPorfolioWithPnlLoading = action.payload;
    },
    setIsWalletPortfolioWithPnlErroring(state, action: PayloadAction<boolean>) {
      state.isWalletPortfolioWithPnlErroring = action.payload;
    },
    setPriceGraphPeriod(state, action: PayloadAction<TokenPriceGraphPeriod>) {
      state.priceGraphPeriod = action.payload;
    },
    setPeriodFilter(state, action: PayloadAction<PeriodFilterBalance>) {
      state.periodFilter = action.payload;
    },
    setPeriodFilterPnl(state, action: PayloadAction<PeriodFilterPnl>) {
      state.periodFilterPnl = action.payload;
    },
    setWalletHistoryGraph(
      state,
      action: PayloadAction<WalletHistory | undefined>
    ) {
      state.walletHistoryGraph = action.payload;
    },
    setIsWalletHistoryGraphLoading(state, action: PayloadAction<boolean>) {
      state.isWalletHistoryGraphLoading = action.payload;
    },
    setIsWalletHistoryGraphErroring(state, action: PayloadAction<boolean>) {
      state.isWalletHistoryGraphErroring = action.payload;
    },
    setTopTokenUnrealizedPnL(
      state,
      action: PayloadAction<WalletHistory | undefined>
    ) {
      state.topTokenUnrealizedPnL = action.payload;
    },
    setIsTopTokenUnrealizedPnLLoading(state, action: PayloadAction<boolean>) {
      state.isTopTokenUnrealizedPnLLoading = action.payload;
    },
    setIsTopTokenUnrealizedPnLErroring(state, action: PayloadAction<boolean>) {
      state.isTopTokenUnrealizedPnLErroring = action.payload;
    },
    setSelectedBalanceOrPnl(state, action: PayloadAction<'balance' | 'pnl'>) {
      state.selectedBalanceOrPnl = action.payload;
    },
    setIsRefreshAll(state, action: PayloadAction<boolean>) {
      state.isRefreshAll = action.payload;
    },
    setIsReceiveModalOpen(state, action: PayloadAction<boolean>) {
      state.isReceiveModalOpen = action.payload;
    },
  },
});

export const {
  setWalletPortfolio,
  setIsWalletPorfolioLoading,
  setIsWalletPortfolioErroring,
  setWalletPortfolioWithPnl,
  setIsWalletPortfolioWithPnlLoading,
  setIsWalletPortfolioWithPnlErroring,
  setPriceGraphPeriod,
  setPeriodFilter,
  setPeriodFilterPnl,
  setWalletHistoryGraph,
  setIsWalletHistoryGraphLoading,
  setIsWalletHistoryGraphErroring,
  setTopTokenUnrealizedPnL,
  setIsTopTokenUnrealizedPnLLoading,
  setIsTopTokenUnrealizedPnLErroring,
  setSelectedBalanceOrPnl,
  setIsRefreshAll,
  setIsReceiveModalOpen,
} = walletPortfolioSlice.actions;

export default walletPortfolioSlice;
