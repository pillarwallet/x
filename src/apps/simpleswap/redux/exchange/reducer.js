import { createReducer } from '@reduxjs/toolkit';

import * as actions from './actions';

const initialState = {
  allCurrenciesList: {
    data: null,
    isLoading: true,
  },
  filter: {
    data: null,
    isLoading: true,
  },
  fixedCurrencies: null,
  allPairsList: null,
  fixedPairs: null,
  fiat: null,
  popularCurrencies: null,
  popularFiat: null,
  exchangesList: [],
};

const reducer = createReducer(initialState, (builder) => {
  builder.addCase(actions.setAllCurrencies, (state, action) => {
    const { data, isLoading } = action.payload;
    if (isLoading)
      return {
        ...state,
        allCurrenciesList: { data: null, isLoading: true },
      };

    const mapCoins = data.filter((coinItem) => !coinItem.isFiat);
    return {
      ...state,
      allCurrenciesList: { data: mapCoins, isLoading: false },
    };
  })
    .addCase(actions.setAllPairs, (state, action) => {
      const { data } = action.payload;

      return {
        ...state,
        allPairsList: data,
      };
    })
    .addCase(actions.setFixedCurrencies, (state, action) => {
      const { data } = action.payload;

      return {
        ...state,
        fixedCurrencies: data,
      };
    })
    .addCase(actions.setFixedPairs, (state, action) => {
      const { data: fixedPairs } = action.payload;

      return {
        ...state,
        fixedPairs,
      };
    })
    .addCase(actions.setPopularCurrencies, (state, action) => {
      const { popularCurrencies } = action.payload;

      return {
        ...state,
        popularCurrencies,
      };
    })
    .addCase(actions.setExchangeInfo, (state, action) => {
      const { data, exchangeId } = action.payload;

      const { exchangesList } = state;

      if (data.error) {
        return state;
      }

      const newExchangesList = exchangesList.filter((i) => i.id !== exchangeId);
      newExchangesList.push(data.data);

      return {
        ...state,
        exchangesList: newExchangesList,
      };
    })
})


export default reducer;
