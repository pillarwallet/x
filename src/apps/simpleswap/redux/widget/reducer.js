/* eslint-disable no-param-reassign */
import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';
import axios from '../../helpers/axios';
import {widgetInfo} from '../../constants/widgetInfo';

const initialState = {
  exchangeInfo: null,
};

export const GET_WIDGET_EXCHANGE_INFO = 'widget/GET_WIDGET_EXCHANGE_INFO';

// eslint-disable-next-line consistent-return
export const getWidgetExchangeInfo = createAsyncThunk(
  GET_WIDGET_EXCHANGE_INFO,
  async ({ exchangeId }) => {
      const headers = {
        'x-api-key': widgetInfo.apiKey,
      };
      const response = await axios.get(`/exchanges/${exchangeId}`, {
        headers,
      });

      return response.data.result;
  },
);

export const widgetSlice = createSlice({
  name: 'widget',
  initialState,
  reducers: {
    resetWidgetExchangeInfo: (state) => {
      state.exchangeInfo = null;
    },
  },
  extraReducers: (builder) => {
    builder
      .addCase(getWidgetExchangeInfo.fulfilled, (state, { payload }) => {
        state.exchangeInfo = payload;
      })
  },
});

export const { resetWidgetExchangeInfo } = widgetSlice.actions;

export default widgetSlice.reducer;
