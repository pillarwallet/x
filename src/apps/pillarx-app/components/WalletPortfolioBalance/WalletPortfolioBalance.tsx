import { useCallback, useMemo } from 'react';
import { TbTriangleFilled } from 'react-icons/tb';

// types
import { PortfolioData, WalletHistory } from '../../../../types/api';

// utils

// reducer
import { useAppDispatch, useAppSelector } from '../../hooks/useReducerHooks';
import { setIsRefreshAll } from '../../reducer/WalletPortfolioSlice';

// images
import RefreshIcon from '../../images/refresh-button.png';
import WalletPortfolioIcon from '../../images/wallet-portfolio-icon.png';

// components
import SkeletonLoader from '../../../../components/SkeletonLoader';
import Body from '../Typography/Body';
import BodySmall from '../Typography/BodySmall';

const WalletPortfolioBalance = () => {
  const dispatch = useAppDispatch();
  const walletPortfolio = useAppSelector(
    (state) =>
      state.walletPortfolio.walletPortfolio as PortfolioData | undefined
  );
  const topTokenUnrealizedPnL = useAppSelector(
    (state) =>
      state.walletPortfolio.topTokenUnrealizedPnL as WalletHistory | undefined
  );
  const isWalletPortfolioLoading = useAppSelector(
    (state) => state.walletPortfolio.isWalletPortfolioLoading as boolean
  );
  const isWalletPortfolioWithPnlLoading = useAppSelector(
    (state) => state.walletPortfolio.isWalletPortfolioWithPnlLoading as boolean
  );
  const isWalletHistoryGraphLoading = useAppSelector(
    (state) => state.walletPortfolio.isWalletHistoryGraphLoading as boolean
  );
  const isTopTokenUnrealizedPnLLoading = useAppSelector(
    (state) => state.walletPortfolio.isTopTokenUnrealizedPnLLoading as boolean
  );

  const isAnyDataFetching =
    isWalletPortfolioLoading ||
    isWalletPortfolioWithPnlLoading ||
    isWalletHistoryGraphLoading ||
    isTopTokenUnrealizedPnLLoading;

  const balanceChange = useMemo(() => {
    if (!topTokenUnrealizedPnL) return undefined;

    const sorted = [...topTokenUnrealizedPnL.balance_history].sort(
      (a, b) => a[0] - b[0]
    );
    const first = sorted[0][1];
    const last = sorted[sorted.length - 1][1];
    const usdValue = last - first;
    const percentageValue = first !== 0 ? (usdValue / first) * 100 : 0;

    return { usdValue, percentageValue };
  }, [topTokenUnrealizedPnL]);

  const getUsdChangeText = useCallback(() => {
    if (
      !balanceChange ||
      !balanceChange.usdValue ||
      balanceChange.usdValue === 0
    )
      return '$0.00';
    const absValue = Math.abs(balanceChange.usdValue);

    return balanceChange.usdValue > 0
      ? `+$${absValue.toFixed(2)}`
      : `-$${absValue.toFixed(2)}`;
  }, [balanceChange]);

  return (
    <div className="flex w-full justify-between">
      <div className="flex flex-col gap-4">
        <div className="flex gap-1.5">
          <img
            src={WalletPortfolioIcon}
            alt="wallet-portfolio-icon"
            className="w-7 h-5"
          />
          <Body>My portfolio</Body>
        </div>
        {isWalletPortfolioLoading || !walletPortfolio ? (
          <SkeletonLoader $height="45px" $width="150px" $radius="10px" />
        ) : (
          <div className="flex flex-wrap gap-1.5 items-baseline">
            <p
              className={`text-[30px] font-semibold ${walletPortfolio.total_wallet_balance > 0 ? 'text-white' : 'text-white text-opacity-50'}`}
            >
              $
              {walletPortfolio.total_wallet_balance > 0
                ? walletPortfolio?.total_wallet_balance.toFixed(2)
                : '0.00'}
            </p>
            <div className="flex tablet:flex-wrap mobile:flex-wrap gap-1.5 items-center">
              {!balanceChange ? null : (
                <Body
                  className={`text-base font-normal ${balanceChange.usdValue > 0 && 'text-market_row_green'} ${balanceChange.usdValue < 0 && 'text-percentage_red'} ${getUsdChangeText() === '$0.00' && 'text-white text-opacity-50'} ${balanceChange.usdValue === 0 && 'text-white text-opacity-50 bg-white/[.1]'}`}
                >
                  {getUsdChangeText()}
                </Body>
              )}
              {!balanceChange ? null : (
                <div
                  className={`flex gap-1 items-center px-1 rounded ${balanceChange.percentageValue > 0 && 'text-market_row_green bg-market_row_green/[.1]'} ${balanceChange.percentageValue < 0 && 'text-percentage_red bg-percentage_red/[.1]'} ${balanceChange.percentageValue === 0 && 'text-white text-opacity-50 bg-white/[.1]'}`}
                >
                  {balanceChange.percentageValue === 0 ? null : (
                    <TbTriangleFilled
                      size={6}
                      color={
                        balanceChange.percentageValue > 0
                          ? '#5CFF93'
                          : '#FF366C'
                      }
                      style={{
                        transform:
                          balanceChange.percentageValue < 0
                            ? 'rotate(180deg)'
                            : 'none',
                      }}
                    />
                  )}

                  <BodySmall className="font-normal">
                    {balanceChange.percentageValue !== 0
                      ? Math.abs(balanceChange.percentageValue).toFixed(2)
                      : '0.00'}
                    %
                  </BodySmall>
                </div>
              )}
              {!balanceChange ? null : (
                <BodySmall
                  className={`font-normal px-1 rounded ${balanceChange.percentageValue === 0 ? 'text-white text-opacity-50 bg-white/[.1]' : 'text-purple_medium bg-purple_medium/[.1]'} `}
                >
                  24h
                </BodySmall>
              )}
            </div>
          </div>
        )}
      </div>
      <div
        className={`flex w-fit h-fit items-center justify-center rounded-[10px] border-x-2 border-t-2 border-b-4 border-[#121116] ${isAnyDataFetching ? 'cursor-not-allowed opacity-50' : 'cursor-pointer'}`}
        onClick={() => {
          if (!isAnyDataFetching) {
            dispatch(setIsRefreshAll(true));
          }
        }}
      >
        <img src={RefreshIcon} alt="refresh-button" className="w-9 h-[34px]" />
      </div>
    </div>
  );
};

export default WalletPortfolioBalance;
