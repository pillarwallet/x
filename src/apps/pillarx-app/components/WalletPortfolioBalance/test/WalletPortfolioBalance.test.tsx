/* eslint-disable @typescript-eslint/no-explicit-any */
import { fireEvent, render, screen } from '@testing-library/react';

// types
import type { Mock } from 'vitest';

// reducer
import {
  useAppDispatch as useAppDispatchMock,
  useAppSelector as useAppSelectorMock,
} from '../../../hooks/useReducerHooks';
import { setIsRefreshAll } from '../../../reducer/WalletPortfolioSlice';

// components
import WalletPortfolioBalance from '../WalletPortfolioBalance';

vi.mock('../../../hooks/useReducerHooks');
vi.mock('../../../../../components/SkeletonLoader', () => ({
  __esModule: true,
  default: function SkeletonLoader() {
    return <div data-testid="skeleton-loader">Loading...</div>;
  },
}));
vi.mock('../../../images/refresh-button.png', () => {
  return {
    __esModule: true,
    default: 'refresh-button.png',
  };
});
vi.mock('../../../images/wallet-portfolio-icon.png', () => {
  return {
    __esModule: true,
    default: 'wallet-portfolio-icon.png',
  };
});

const mockDispatch = vi.fn();
(useAppDispatchMock as unknown as Mock).mockReturnValue(mockDispatch);

describe('WalletPortfolioBalance', () => {
  beforeEach(() => {
    vi.clearAllMocks();
  });

  const baseSelectorState = {
    walletPortfolio: {
      isWalletPortfolioLoading: false,
      isWalletPortfolioWithPnlLoading: false,
      isWalletHistoryGraphLoading: false,
      isTopTokenUnrealizedPnLLoading: false,
    },
  };

  const renderWithSelector = ({
    walletPortfolio,
    topTokenUnrealizedPnL,
    overrides = {},
  }: {
    walletPortfolio?: any;
    topTokenUnrealizedPnL?: any;
    overrides?: Partial<(typeof baseSelectorState)['walletPortfolio']>;
  }) => {
    (useAppSelectorMock as unknown as Mock).mockImplementation((selectorFn) =>
      selectorFn({
        walletPortfolio: {
          walletPortfolio,
          topTokenUnrealizedPnL,
          ...baseSelectorState.walletPortfolio,
          ...overrides,
        },
      })
    );

    render(<WalletPortfolioBalance />);
  };

  it('renders correctly and matches snapshot', () => {
    const tree = render(<WalletPortfolioBalance />);
    expect(tree).toMatchSnapshot();
  });

  it('shows skeleton when loading', () => {
    renderWithSelector({
      walletPortfolio: undefined,
      overrides: { isWalletPortfolioLoading: true },
    });

    expect(screen.getByTestId('skeleton-loader')).toBeInTheDocument();
  });

  it('renders 0 balance when wallet is empty', () => {
    renderWithSelector({
      walletPortfolio: { total_wallet_balance: 0 },
      topTokenUnrealizedPnL: undefined,
    });

    expect(screen.getByText('$0.00')).toBeInTheDocument();
    expect(screen.getByAltText('wallet-portfolio-icon')).toBeInTheDocument();
  });

  it('renders balance with positive change', () => {
    renderWithSelector({
      walletPortfolio: { total_wallet_balance: 1000 },
      topTokenUnrealizedPnL: {
        balance_history: [
          [1, 500],
          [2, 1000],
        ],
      },
    });

    expect(screen.getByText('$1000.00')).toBeInTheDocument();
    expect(screen.getByText(/\+?\$500/)).toBeInTheDocument();
    expect(screen.getByText('100.00%')).toBeInTheDocument();
    expect(screen.getByText('24h')).toBeInTheDocument();
  });

  it('renders balance with negative change', () => {
    renderWithSelector({
      walletPortfolio: { total_wallet_balance: 1000 },
      topTokenUnrealizedPnL: {
        balance_history: [
          [1, 1000],
          [2, 500],
        ],
      },
    });

    expect(screen.getByText('-$500.00')).toBeInTheDocument();
    expect(screen.getByText('50.00%')).toBeInTheDocument();
  });

  it('renders balance with zero change', () => {
    renderWithSelector({
      walletPortfolio: { total_wallet_balance: 1000 },
      topTokenUnrealizedPnL: {
        balance_history: [
          [1, 1000],
          [2, 1000],
        ],
      },
    });

    expect(screen.queryByText(/\$500/)).not.toBeInTheDocument();
    expect(screen.queryByText(/100.00%/)).not.toBeInTheDocument();
    expect(screen.queryByText('0.00%')).toBeInTheDocument();
    expect(screen.getByText('24h')).toBeInTheDocument();
  });

  it('disables refresh button while loading data', () => {
    renderWithSelector({
      walletPortfolio: { total_wallet_balance: 1000 },
      overrides: {
        isWalletPortfolioLoading: true,
        isWalletPortfolioWithPnlLoading: true,
      },
    });

    const button = screen.getByAltText('refresh-button').parentElement;
    expect(button).toHaveClass('opacity-50');
    fireEvent.click(button!);
    expect(mockDispatch).not.toHaveBeenCalled();
  });

  it('enables refresh button and dispatches action', () => {
    renderWithSelector({
      walletPortfolio: { total_wallet_balance: 1000 },
    });

    const button = screen.getByAltText('refresh-button').parentElement;
    expect(button).not.toHaveClass('opacity-50');
    fireEvent.click(button!);
    expect(mockDispatch).toHaveBeenCalledWith(setIsRefreshAll(true));
  });
});
