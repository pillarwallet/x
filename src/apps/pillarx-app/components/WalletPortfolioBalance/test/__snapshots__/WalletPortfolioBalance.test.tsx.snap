// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`WalletPortfolioBalance > renders correctly and matches snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="flex w-full justify-between"
      >
        <div
          class="flex flex-col gap-4"
        >
          <div
            class="flex gap-1.5"
          >
            <img
              alt="wallet-portfolio-icon"
              class="w-7 h-5"
              src="wallet-portfolio-icon.png"
            />
            <p
              class="text-base font-medium undefined"
            >
              My portfolio
            </p>
          </div>
          <div
            data-testid="skeleton-loader"
          >
            Loading...
          </div>
        </div>
        <div
          class="flex w-fit h-fit items-center justify-center rounded-[10px] border-x-2 border-t-2 border-b-4 border-[#121116] cursor-pointer"
        >
          <img
            alt="refresh-button"
            class="w-9 h-[34px]"
            src="refresh-button.png"
          />
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="flex w-full justify-between"
    >
      <div
        class="flex flex-col gap-4"
      >
        <div
          class="flex gap-1.5"
        >
          <img
            alt="wallet-portfolio-icon"
            class="w-7 h-5"
            src="wallet-portfolio-icon.png"
          />
          <p
            class="text-base font-medium undefined"
          >
            My portfolio
          </p>
        </div>
        <div
          data-testid="skeleton-loader"
        >
          Loading...
        </div>
      </div>
      <div
        class="flex w-fit h-fit items-center justify-center rounded-[10px] border-x-2 border-t-2 border-b-4 border-[#121116] cursor-pointer"
      >
        <img
          alt="refresh-button"
          class="w-9 h-[34px]"
          src="refresh-button.png"
        />
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
